### PR TITLE
feat: create_(random)_proof_many

### PR DESCRIPTION
**Pull request by dignifiedquire**

### CODE REVIEW

File: src/groth16/prover.rs [hunk 1]
```diff
@@ -175,6 +175,306 @@ where
...
+            }
+
+            Ok(prover)
+        })
+        .collect::<Result<Vec<_>, _>>()?;
+
+    let worker = Worker::new();
+    let input_len = provers[0].input_assignment.len();
+    let vk = params.get_vk(input_len)?;
+    let n = provers[0].a.len();
```
Comment (keyvank) [inline src/groth16/prover.rs]: Does that mean all circuits have equal number of constraints?

File: src/groth16/prover.rs [hunk 2]
```diff
@@ -175,6 +175,306 @@ where
...
+    let mut log_d = 0u32;
+    while (1 << log_d) < n {
+        log_d += 1;
+    }
+
+    #[cfg(feature = "gpu")]
+    let lock = gpu::lock()?;
+
+    let mut fft_kern = create_fft_kernel(log_d);
+    let mut multiexp_kern = create_multiexp_kernel();
```
Comment (keyvank) [inline src/groth16/prover.rs]: This should be instantiated, after `fft_kern` is dropped. I'll fix it myself

File: src/groth16/prover.rs (hunk 1 above)
Comment (dignifiedquire) [reply src/groth16/prover.rs]: yes, this method is specifically designed to process `n` circuits that are identical, just with different values

File: src/groth16/tests/mod.rs [hunk 3]
```diff
@@ -378,3 +381,58 @@ fn test_xordemo() {
...
+    )
+    .unwrap();
+
+    assert_eq!(proof_batch[0], proof_single_1);
+    assert_eq!(proof_batch[1], proof_single_2);
+
+    assert!(verify_proof(&pvk, &proof_single_1, &[Fr::one()]).unwrap());
+    assert!(verify_proof(&pvk, &proof_single_2, &[Fr::one()]).unwrap());
+    for proof in &proof_batch {
+        assert!(verify_proof(&pvk, &proof, &[Fr::one()]).unwrap());
```
Comment (keyvank) [inline src/groth16/tests/mod.rs]: Isn't this redundant?

**Comment by keyvank** [comment]
Also, I think its better to call `create_proof_batch` inside `create_proof` instead of having everything implemented twice.

File: src/groth16/tests/mod.rs (hunk 3 above)
Comment (dignifiedquire) [reply src/groth16/tests/mod.rs]: why?

**Pull request merged**